### PR TITLE
Setup maven wrapper

### PR DESCRIPTION
**Pull request by worldtiki**
# Pull Request Template

## Description

Add maven wrapper to this project (mvn version 3.6.1).
Certain libraries (for eg. junit 5) require different configurations depending on the version of maven used. By allowing users to have a fully encapsulated build setup we avoid these issues.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [X] Just a regular mvn clean install

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented on my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] Any dependent changes have been merged and published in downstream modules
- [X] My changes have no bad impacts on performances
- [ ] Any implemented change has been added in the [CHANGELOG.md](./CHANGELOG.md) file

### CODE REVIEW

**Comment by worldtiki** [comment]
Not sure if this needs to be added to the changelog as there were no changes to the library code itself.

**Comment by fborriello** [comment]
> Hi @fborriello!
> 
> This is generated code (https://github.com/takari/maven-wrapper) and I would not recommend changing it.

I know, but this means that every time there is a new maven release and we want to upgrade it, we need re-generate the code from takari project, isn't it?
But, if apply the below changes, the maven version upgrade can be performed by simply update a value in the properties file.
What do you think?

**Comment by fborriello** [comment]
> Hi @fborriello!
> 
> This is generated code (https://github.com/takari/maven-wrapper) and I would not recommend changing it.

I know, but this implies that every time we need to upgrade the maven version, we need to use the takari project to re-generate the maven-wrapper with the latest version.
But, if we apply the suggested changes, the upgrade will be done by simply modifying a value in a properties file. what do you think?

**Comment by worldtiki** [comment]
> > Hi @fborriello!
> > This is generated code (https://github.com/takari/maven-wrapper) and I would not recommend changing it.
> 
> I know, but this implies that every time we need to upgrade the maven version, we need to use the takari project to re-generate the maven-wrapper with the latest version.
> But, if we apply the suggested changes, the upgrade will be done by simply modifying a value in a properties file. what do you think?

I agree it would make the upgrade process a bit simpler, but it would require further changes.
The properties file is handled by org.apache.maven.wrapper.WrapperExecutor which unfortunately does not seem to handle placeholders.

It results in this error:

```
Exception in thread "main" java.lang.RuntimeException: Could not load wrapper properties from '/Users/dmartinsalbuqu/Development/Workspaces/bull/.mvn/wrapper/maven-wrapper.properties'.
        at org.apache.maven.wrapper.WrapperExecutor.<init>(WrapperExecutor.java:72)
        at org.apache.maven.wrapper.WrapperExecutor.forWrapperPropertiesFile(WrapperExecutor.java:56)
        at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:60)
Caused by: java.net.URISyntaxException: Illegal character in path at index 68: https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/${maven.version}/apache-maven-${maven.version}-bin.zip
```

**Comment by fborriello** [comment]
> > > Hi @fborriello!
> > > This is generated code (https://github.com/takari/maven-wrapper) and I would not recommend changing it.
> > 
> > 
> > I know, but this implies that every time we need to upgrade the maven version, we need to use the takari project to re-generate the maven-wrapper with the latest version.
> > But, if we apply the suggested changes, the upgrade will be done by simply modifying a value in a properties file. what do you think?
> 
> I agree it would make the upgrade process a bit simpler, but it would require further changes.
> The properties file is handled by org.apache.maven.wrapper.WrapperExecutor which unfortunately does not seem to handle placeholders.
> 
> It results in this error:
> 
> ```
> Exception in thread "main" java.lang.RuntimeException: Could not load wrapper properties from '/bull/.mvn/wrapper/maven-wrapper.properties'.
>         at org.apache.maven.wrapper.WrapperExecutor.<init>(WrapperExecutor.java:72)
>         at org.apache.maven.wrapper.WrapperExecutor.forWrapperPropertiesFile(WrapperExecutor.java:56)
>         at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:60)
> Caused by: java.net.URISyntaxException: Illegal character in path at index 68: https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/${maven.version}/apache-maven-${maven.version}-bin.zip
> ```

if you don't mind, I would try to figure out a solution for this, do you have an available branch that raises the above error?

**Comment by worldtiki** [comment]
> > > > Hi @fborriello!
> > > > This is generated code (https://github.com/takari/maven-wrapper) and I would not recommend changing it.
> > > 
> > > 
> > > I know, but this implies that every time we need to upgrade the maven version, we need to use the takari project to re-generate the maven-wrapper with the latest version.
> > > But, if we apply the suggested changes, the upgrade will be done by simply modifying a value in a properties file. what do you think?
> > 
> > 
> > I agree it would make the upgrade process a bit simpler, but it would require further changes.
> > The properties file is handled by org.apache.maven.wrapper.WrapperExecutor which unfortunately does not seem to handle placeholders.
> > It results in this error:
> > ```
> > Exception in thread "main" java.lang.RuntimeException: Could not load wrapper properties from '/bull/.mvn/wrapper/maven-wrapper.properties'.
> >         at org.apache.maven.wrapper.WrapperExecutor.<init>(WrapperExecutor.java:72)
> >         at org.apache.maven.wrapper.WrapperExecutor.forWrapperPropertiesFile(WrapperExecutor.java:56)
> >         at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:60)
> > Caused by: java.net.URISyntaxException: Illegal character in path at index 68: https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/${maven.version}/apache-maven-${maven.version}-bin.zip
> > ```
> 
> if you don't mind, I would try to figure out a solution for this, do you have an available branch that raises the above error?

You can replace the contents of the properties file with:

```
maven.version=3.6.1
distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/${maven.version}/apache-maven-${maven.version}-bin.zip
maven.wrapper.version=0.5.5
wrapperUrl=https://repo.maven.apache.org/maven2/io/takari/maven-wrapper/${maven.wrapper.version}/maven-wrapper-${maven.wrapper.version}.jar
```
And try to run maven, for example:
```./mvn -version```

File: .mvn/wrapper/maven-wrapper.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.1/apache-maven-3.6.1-bin.zip
```
[the remark this answers is not in the document]
Comment (fborriello) [reply .mvn/wrapper/maven-wrapper.properties]: You're right, however, I've implemented a working solution that even centralizes the `maven-wrapper` and the maven distribution into the properties file. 
I pushed the solution into the branch: `maven_wrapper` on your repo. If you agree with the changes we can proceed with the PR process.

[the remark this answers is not in the document]
Comment (worldtiki) [reply .mvn/wrapper/maven-wrapper.properties]: Of course! 🙂

**Pull request merged**